File: src/modules/signal-processing/KalmanFilter.ts

/**
 * Implementación de Filtro Kalman para procesamiento de señal
 */
export class KalmanFilter {
  private R: number = 0.05; // Reducido de 0.09 a 0.05 - Menor varianza de medición (más confianza en señal original)
  private Q: number = 1.2;  // Aumentado de 0.8 a 1.2 - Mayor varianza del proceso (permite cambios mucho más bruscos)
  private P: number = 1;    // Covarianza del error estimado
  private X: number = 0;    // Estado estimado
  private K: number = 0;    // Ganancia de Kalman

  filter(measurement: number): number {
    // Predicción
    this.P = this.P + this.Q;
    
    // Actualización
    this.K = this.P / (this.P + this.R);
    this.X = this.X + this.K * (measurement - this.X);
    this.P = (1 - this.K) * this.P;
    
    return this.X;
  }

  reset() {
    this.X = 0;
    this.P = 1;
  }
}

File: src/modules/signal-processing/KalmanFilter.kt
package modules.signal_processing

class KalmanFilter {
    // Varianza de la medición (ruido del sensor)
    private var R: Double = 0.01 
    // Varianza del proceso
    private var Q: Double = 0.1  
    // Covarianza del error estimado
    private var P: Double = 1.0    
    // Estado estimado
    private var X: Double = 0.0    
    // Ganancia de Kalman (no es necesario inicializarla aquí si se calcula primero en filter)
    // private var K: Double = 0.0 

    fun filter(measurement: Double): Double {
        // Predicción
        // X = X  (estado no cambia en esta simple predicción)
        P += Q

        // Actualización
        val K = P / (P + R) // Ganancia de Kalman
        X += K * (measurement - X)
        P *= (1 - K)

        return X
    }

    fun reset() {
        // Restablecer a los valores iniciales o predeterminados
        R = 0.01
        Q = 0.1
        P = 1.0
        X = 0.0
        // K se recalcula en cada paso de filter, así que no necesita resetearse explícitamente aquí.
    }
} 